## Data Cleaning
As we all know when data is collected, it most likely comes in an unordered manner. With inconsistency and incompleteness.  
So, it is our job to address this data and process it optimally, so it could be used as information for insightful analysis or to train a model. 

In [1]:
# Import package(s) for data exploration.
import pandas as pd

# Ignoring all unnecessary warnings.
pd.set_option('mode.chained_assignment', None)

In [2]:
# Load the data into a pandas dataframe.
df = pd.read_csv(r"C:\Users\ifunanyaScript\Everything\House_price_prediction_repository\Bengaluru_House_Data.csv")
df.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [3]:
#  rows, columns
df.shape

(13320, 9)

The point of data cleaning as the name suggest is to get rid of unnecessary data and keep important information. 
For example, we can start by examining the first column "area type", to see how much effect it has on the data.

In [4]:
# Number of houses in each area type. 
df.groupby('area_type')['area_type'].agg('count')

area_type
Built-up  Area          2418
Carpet  Area              87
Plot  Area              2025
Super built-up  Area    8790
Name: area_type, dtype: int64

There are only 4 area types and nothing seems special about their distribution, except the carpet area type with only 87 house counts.  
Let's see if there's something going on there...

In [5]:
# The dataframe where area type is Carpet Area.
df[df["area_type"]=="Carpet  Area"].head(15)

area_type   availability                  location       size  \
297   Carpet  Area  Ready To Move         Maruthi Sevanagar      2 BHK   
340   Carpet  Area  Ready To Move          Muneshwara Nagar      2 BHK   
352   Carpet  Area  Ready To Move              Sampigehalli      3 BHK   
594   Carpet  Area  Ready To Move              Sanjay nagar      2 BHK   
622   Carpet  Area  Ready To Move         Bannerghatta Road      3 BHK   
629   Carpet  Area         18-Jul                 Hosahalli      3 BHK   
642   Carpet  Area  Ready To Move            Chikkalasandra      2 BHK   
1020  Carpet  Area  Ready To Move            Weavers Colony      1 BHK   
1033  Carpet  Area  Ready To Move                Nagasandra  6 Bedroom   
1181  Carpet  Area  Ready To Move                   Iggalur      2 BHK   
1384  Carpet  Area  Ready To Move  Electronics City Phase 1      1 BHK   
1420  Carpet  Area  Ready To Move            Vinayaka Nagar      3 BHK   
1484  Carpet  Area         18-Mar                    Hebbal      2 BHK   
1499  Carpet  Area         18-Dec               Bhoganhalli      2 BHK   
1605  Carpet  Area  Ready To Move               Kumara Park      3 BHK   

      society       total_sqft  bath  balcony   price  
297   SMikaay              950   2.0      2.0   47.00  
340       NaN             1230   2.0      3.0   48.00  
352   Puens G             1592   3.0      2.0   75.00  
594       NaN             1100   2.0      2.0   98.00  
622       NaN             1445   3.0      2.0   95.00  
629   NSezeBr             1375   3.0      2.0   70.00  
642       NaN              875   2.0      3.0   52.80  
1020      NaN               15   1.0      0.0   30.00  
1033      NaN             1500   5.0      1.0  130.00  
1181  VaoldMa              910   2.0      1.0   30.50  
1384      NaN              785   1.0      1.0   55.00  
1420  SVon hn             1535   3.0      1.0   95.00  
1484  Arcia S  547.34 - 827.31   2.0      0.0   42.72  
1499  Rosha I            804.1   2.0      1.0   69.09  
1605      NaN             2200   4.0      NaN  350.00

Apparently the prices are still scattered accross.  
We can effectively say the area type feature is redundant.
<br>
<br>
We can proceed to the next column/feature and examine that as well...

In [6]:
df.groupby('availability')['availability'].agg('count')

availability
14-Jul                      1
14-Nov                      1
15-Aug                      1
15-Dec                      1
15-Jun                      1
                        ...  
22-Mar                      3
22-May                     10
22-Nov                      2
Immediate Possession       16
Ready To Move           10581
Name: availability, Length: 81, dtype: int64

This column seems redundant so I will forego this.  
However, in a situation where one isn't sure of foregoing a particular feature, one can always consult a senior or more experienced colleague in the field.
<br>
<br>
Let's examine the "society" column...

In [7]:
df.groupby('society')['society'].agg('count')

society
3Codeli    2
7 ise P    1
A idse     2
A rtsai    1
ACersd     1
          ..
Zonce E    2
Zostaa     3
i1ncyRe    1
i1odsne    1
i1rtsCo    3
Name: society, Length: 2688, dtype: int64

There are 2688 different society types, and the houses per society is of a minuscule magnitude.  
We can effectively label this as a redundant feature.
<br>
<br>
The balcony column seems like a very important feature. However let's take a look...

In [8]:
df.groupby('balcony')['balcony'].agg('count')

balcony
0.0    1029
1.0    4897
2.0    5113
3.0    1672
Name: balcony, dtype: int64

We can take a look at houses that have 3 balconies, let's see how that affects the cost of the house("price").

In [9]:
df[df["balcony"]==3].head(15)

area_type   availability               location       size  \
1              Plot  Area  Ready To Move       Chikka Tirupathi  4 Bedroom   
2          Built-up  Area  Ready To Move            Uttarahalli      3 BHK   
11             Plot  Area  Ready To Move             Whitefield  4 Bedroom   
17   Super built-up  Area  Ready To Move  Raja Rajeshwari Nagar      3 BHK   
33   Super built-up  Area  Ready To Move  Raja Rajeshwari Nagar      3 BHK   
54   Super built-up  Area  Ready To Move              Yelahanka      3 BHK   
62             Plot  Area  Ready To Move             Whitefield  4 Bedroom   
66   Super built-up  Area         18-Apr                 Gunjur      2 BHK   
76   Super built-up  Area  Ready To Move        Kalena Agrahara      2 BHK   
77         Built-up  Area  Ready To Move     Cholanayakanahalli      3 BHK   
79         Built-up  Area  Ready To Move            ISRO Layout  6 Bedroom   
89             Plot  Area  Ready To Move           Rajaji Nagar  6 Bedroom   
94   Super built-up  Area         17-Aug            Thanisandra      2 BHK   
100  Super built-up  Area  Ready To Move           Yeshwanthpur      3 BHK   
110  Super built-up  Area  Ready To Move     Lakshmiamma Garden      3 BHK   

     society total_sqft  bath  balcony   price  
1    Theanmp       2600   5.0      3.0  120.00  
2        NaN       1440   2.0      3.0   62.00  
11   Prrry M       2785   5.0      3.0  295.00  
17   GrrvaGr       1540   3.0      3.0   60.00  
33   GrrvaGr       1693   3.0      3.0   57.39  
54   NCowsar       1600   2.0      3.0   75.00  
62   Chranya       5700   5.0      3.0  650.00  
66   BMineis       1140   2.0      3.0   43.00  
76       NaN       1150   2.0      3.0   40.00  
77   Viony H       2511   3.0      3.0  205.00  
79       NaN       4400   6.0      3.0  250.00  
89       NaN        710   6.0      3.0  160.00  
94   Bhmesy        1296   2.0      3.0   80.00  
100  IBityin       2502   3.0      3.0  138.00  
110      NaN       1800   3.0      3.0  120.00

Looking at this frame, it clear that balcony does not affect the price, we can see houses that cost 650Lakh and 40Lakh both possessing 3 balconies, yet extremely varying prices.  
This goes to show that balcony could as well be regarded as a redundant feature.

We have spotted 4 __features/columns__ that are redundant, we can reduce dimensionality by dropping these columns.  
Sometimes, one might want to consult their real estate manager before foregoing these columns.  
However, from the above analysis it is clear that those four columns need to be dropped.

In [10]:
# Get rid of unnecessary columns.
df1 = df.drop(['area_type', 'availability', 'society', 'balcony'], axis=1)
df1.head()

location       size total_sqft  bath   price
0  Electronic City Phase II      2 BHK       1056   2.0   39.07
1          Chikka Tirupathi  4 Bedroom       2600   5.0  120.00
2               Uttarahalli      3 BHK       1440   2.0   62.00
3        Lingadheeranahalli      3 BHK       1521   3.0   95.00
4                  Kothanur      2 BHK       1200   2.0   51.00

Now we have a more concise dataframe, we can proceed.  
When we collect data, there is always a chance that there will be holes in the data, i.e missing datapoints.  
These missing datapoints are represented as NaN in the dataframe. These missing datapoints could arise under certain circumstances but we do not need such holes in our data.  
Let's see how many holes are in each column of our data.

In [11]:
# A count of the null data points in each column.
df1.isnull().sum()

location       1
size          16
total_sqft     0
bath          73
price          0
dtype: int64

We have a substantial volume of data, these missing points will be inconsequential.  
I will drop any __entry(row)__ that has a NaN value.  
Anyways, you can interpolate however you please if you do not wish to reduce you data volume.

In [12]:
# Drop every entry with NaN; row-wise.
df2 = df1.dropna()

print(df2.shape)
df2.isnull().sum()

(13246, 5)


location      0
size          0
total_sqft    0
bath          0
price         0
dtype: int64

Now, looking at the size column; the values in this columns are of two different forms, __"Bedroom"__ and __"BHK"__.  

In [13]:
df2['size'].unique()

array(['2 BHK', '4 Bedroom', '3 BHK', '4 BHK', '6 Bedroom', '3 Bedroom',
       '1 BHK', '1 RK', '1 Bedroom', '8 Bedroom', '2 Bedroom',
       '7 Bedroom', '5 BHK', '7 BHK', '6 BHK', '5 Bedroom', '11 BHK',
       '9 BHK', '9 Bedroom', '27 BHK', '10 Bedroom', '11 Bedroom',
       '10 BHK', '19 BHK', '16 BHK', '43 Bedroom', '14 BHK', '8 BHK',
       '12 Bedroom', '13 BHK', '18 Bedroom'], dtype=object)

If you do a little check on what BHK means, it stands for Bedroom, Hall and Kitchen.  
Hence, we can refer to all of the values in size as number of bedrooms.  
We will create a new __feature/column__, bedrooms and its values will be the corresponding number from size column.

In [14]:
# New column that takes the number from size column, e.g '2 BHK' will be '2' in the new column.
df2['bedrooms'] = df2['size'].apply(lambda x: int(x.split(' ')[0]))

In [15]:
df2 = df2.drop(['size'], axis=1)
df2.head()

location total_sqft  bath   price  bedrooms
0  Electronic City Phase II       1056   2.0   39.07         2
1          Chikka Tirupathi       2600   5.0  120.00         4
2               Uttarahalli       1440   2.0   62.00         3
3        Lingadheeranahalli       1521   3.0   95.00         3
4                  Kothanur       1200   2.0   51.00         2

Now that we have our new columns, let's examine it.  
We'll take a look at the different numbers of bedrooms accross the entire dataset.

In [16]:
df2.bedrooms.unique()

array([ 2,  4,  3,  6,  1,  8,  7,  5, 11,  9, 27, 10, 19, 16, 43, 14, 12,
       13, 18], dtype=int64)

The values vary well. However, there are some alarming number od bedrooms here, like 43 bedrooms.  
Let's have a look at our dataframe where number of bedrooms is greater than 15.

In [17]:
# The entries in the dataset where bedrooms is > 15.
df2[df2.bedrooms>15]

location total_sqft  bath  price  bedrooms
1718       2Electronic City Phase II       8000  27.0  230.0        27
3379                  1Hanuman Nagar       2000  16.0  490.0        19
3609   Koramangala Industrial Layout      10000  16.0  550.0        16
4684                     Munnekollal       2400  40.0  660.0        43
11559                  1Kasavanhalli       1200  18.0  200.0        18

There's definitely an error there. A house with 43 bedrooms cannot be of 2400sqft, neither can a house with 18 bedrooms be of 1200 sqft. The area should be a lot bigger. That is and issue we'll address.  
Let's have a look at the __total_sqft__ feature.

In [18]:
df2.total_sqft.unique() 

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

The values in this column seems to be in different forms. There are range values (0000-1111), and other units of length measurement. This is also another problem that we have to address.  
Range values can be averaged to a single value. Perhaps, that is a good way to solve this issue.  
First, let's see how many entries in our data have a ranged value in the total_sqft column. 

In [19]:
# Returns True if a value is float, otherwise returns False
def convertToFloat(x):
    try:
        float(x)
    except:
        return False
    return True

In [20]:
# The entries in the dataset where the total_sqft is not a float; using the (~).
df2[~df2['total_sqft'].apply(convertToFloat)]

location   total_sqft  bath    price  bedrooms
30              Yelahanka  2100 - 2850   4.0  186.000         4
122                Hebbal  3067 - 8156   4.0  477.000         4
137    8th Phase JP Nagar  1042 - 1105   2.0   54.005         2
165              Sarjapur  1145 - 1340   2.0   43.490         2
188              KR Puram  1015 - 1540   2.0   56.800         2
...                   ...          ...   ...      ...       ...
12975          Whitefield   850 - 1060   2.0   38.190         2
12990      Talaghattapura  1804 - 2273   3.0  122.000         3
13059              Harlur  1200 - 1470   2.0   72.760         2
13265               Hoodi  1133 - 1384   2.0   59.135         2
13299          Whitefield  2830 - 2882   5.0  154.500         4

[190 rows x 5 columns]

190 entries.  
As I said earlier we can take the average of these range values and use them instead.

In [21]:
# Takes a range and returns the average of the values.
def convertRangeToFloat(x):
    numbers = x.split(' - ')
    if len(numbers)==2:
        return (float(numbers[0]) + float(numbers[1]))/len(numbers)
    try:
        return float(x)
    except:
        # Return None if the value cannot be converted to float.
        return None

Let's test and compare the function.

In [22]:
convertRangeToFloat('2334 - 7890'), (2334+7890)/2

(5112.0, 5112.0)

I'll create a new dataframe and apply the above function to the total_sqft column.

In [23]:
df3 = df2.copy()
df3.total_sqft = df3.total_sqft.apply(convertRangeToFloat)

# An entry formerly with a range as total_sqft.
df3.loc[30]

location      Yelahanka
total_sqft       2475.0
bath                4.0
price             186.0
bedrooms              4
Name: 30, dtype: object

Perfect!

In [24]:
df3.head()

location  total_sqft  bath   price  bedrooms
0  Electronic City Phase II      1056.0   2.0   39.07         2
1          Chikka Tirupathi      2600.0   5.0  120.00         4
2               Uttarahalli      1440.0   2.0   62.00         3
3        Lingadheeranahalli      1521.0   3.0   95.00         3
4                  Kothanur      1200.0   2.0   51.00         2

We've cleaned our data to a good extent, we can move onto the next step.  
However, there's still more to do, the data is not at an optimal form yet.  
We'll save this clean dataframe with jupyter store magic and continue in the next notebook... __Feature Engineering__.

In [25]:
%store df3

Stored 'df3' (DataFrame)


In [26]:
# ifunanyaScript